src/corehost/cli/fx_reference.cpp

```diff
@@ -23,39 +23,60 @@ bool fx_reference_t::is_roll_forward_compatible(const fx_ver_t& other) const
+        "Code assumes ordering of roll-forward options from least restrictive to most restrictive");
+    static_assert(
+        roll_forward_option::Major > roll_forward_option::LatestMinor,
+        "Code assumes ordering of roll-forward options from least restrictive to most restrictive");
     if (get_fx_version_number().get_minor() != other.get_minor()
-        && roll_fwd_on_no_candidate_fx != roll_fwd_on_no_candidate_fx_option::major
-        && roll_fwd_on_no_candidate_fx != roll_fwd_on_no_candidate_fx_option::minor)
+        && roll_forward < roll_forward_option::Minor)
     {
         return false;
     }
 
     // Verify patch roll forward
     // We do not distinguish here whether a previous framework reference found a patch version based on:
     //  - initial reference matching a patch version,
-    //  - or roll_fwd_on_no_candidate_fx_option=major\minor finding a compatible patch version as initial framework,
+    //  - or roll_forward=major\minor finding a compatible patch version as initial framework,
     //  - or applyPatches=true finding a newer patch version
-    if (get_fx_version_number().get_patch() != other.get_patch()
-        && patch_roll_fwd == false
-        && roll_fwd_on_no_candidate_fx != roll_fwd_on_no_candidate_fx_option::major
-        && roll_fwd_on_no_candidate_fx != roll_fwd_on_no_candidate_fx_option::minor)
+    static_assert(
+        roll_forward_option::Minor > roll_forward_option::LatestPatch,
+        "Code assumes ordering of roll-forward options from least restrictive to most restrictive");
+    static_assert(
+        roll_forward_option::LatestPatch > roll_forward_option::Disable,
+        "Code assumes ordering of roll-forward options from least restrictive to most restrictive");
+
+    if (roll_forward == roll_forward_option::Disable)
     {
+        // In this case no roll-forward is allowed, at all.
+        // And we know the versions are different since we compared 100% equality above, so they're not compatible.
+        // The versions could differ in patch or pre-release, in both cases they're not compatible.
         return false;
     }
 
-    // Release cannot roll forward to pre-release
-    if (!get_fx_version_number().is_prerelease() && other.is_prerelease())
+    if (roll_forward == roll_forward_option::LatestPatch &&
+        apply_patches == false &&
+        (!get_fx_version_number().is_prerelease() || get_fx_version_number().get_patch() != other.get_patch()))
     {
+        // If LatestPatch and applyPatches=false, it is almost the same as Disable.
+        // Special case is pre-release and applyPatches=false, to maintain backward compatibility applyPatches is ignored
+        // on pre-release version for roll forward over pre-release. So in that case allow roll forward if patch versions are the same.
         return false;
     }
 
+    // It's OK to roll forward from release to pre-release and vice versa
```

Comment:
Hmmm?  I thought the algorithm only allowed roll forward to prerelease if one version was specifically requesting prerelease.

Reply:
This is a bit complicated. The function here is only used to reconcile framework references. It is NOT used (anymore) to compare framework reference to a version on disk.
As such this is about what happens if app and framework reference different versions of the same framework.
In general release and pre-release are allowed to roll forward to each other (in 2.* release was allowed to roll forward to pre-release if necessary. Pre-release was not allowed to roll to release, but this is an intentional change in 3.0 - it's relaxing).
The only case where the spec talks about restrictions in this area is that release should prefer to roll forward to release, and only roll to pre-release if no release is available.
So yes, we're tiny little bit violating the spec here since in theory if there are enough framework references in a chain we could get to a situation where the app wants a release, one of the frameworks wants a higher pre-release, and we pick a pre-release even if higher release is available (since we reconcile to the pre-release reference from the framework).
It's not completely wrong, since the app still allows to roll to pre-release in theory - it's basically a gray area.

To implement the behavior were release "wins" would be much harder and we would basically need another pass to reconcile this special case. Probably not worth it.

I'll add a comment into the code describing this potentially weird case, just for completeness.

Reply:
I'll look into this some more... maybe this can be done perfectly without much effort.

Reply:
I now fixed the problem with "prefer release" across reference merging. It means we store more state on framework reference, but it sort of cleans up the code. See the commit https://github.com/dotnet/core-setup/pull/5691/commits/811078839ac416da41b8502f833084ab7eb39ea3.